pmlc/rt/runtime_registry.h

```diff
@@ -0,0 +1,82 @@
+// Copyright 2020, Intel Corporation
+
+#pragma once
+
+#include <functional>
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "llvm/ADT/StringRef.h"
+
+#include "pmlc/rt/runtime.h"
+
+namespace pmlc::rt {
+
+// Factory is used to instantiate a Runtime.
+//
+// Factory is invoked on-demand (typically well after process initialization) to
+// instantiate a runtime, either as a specific request by the PlaidML caller or
+// for general device enumeration.  The registry is responsible for arranging
+// runtimes to be instantiated exactly once; Factory implementations do not need
+// to memoize their results.
+using Factory = std::function<std::shared_ptr<Runtime>()>;
+
+// Loaders instantiate Factories.
+//
+// PlaidML Loaders are (typically) simple, stateless functions, automatically
+// registered via LoaderRegistration when their containing library is loaded
+// into the process.
+//
+// Loaders are invoked on-demand (typically well after process initialization)
+// to determine the runtime factories they're able to provide.
+using Loader = std::function<std::unordered_map<std::string, Factory>()>;
```

Comment:
I'm still having trouble with the idea of having a Factory, a Loader, a Runtime, and a Device. To me all we need is a Runtime, and a runtime takes IR along with a device_id to produce an Executable. Runtimes are registered at startup using static initialization, making it a linker problem. Then for initialization, each registered runtime gets initialized during `plaidml::exec::init()`.

Reply:
We need to parse device IDs anyway in order to figure out the correct runtime.

But: if you'd like to have `Runtime` create an `Executable`, and pass in an ordinal to indicate the logical device entity to use, we can definitely do that.  It only becomes an anti-pattern when we try to do anything else with the logical device entity, e.g. querying its properties -- the set of things you can do with a number aren't the same set of things you can do with a `Device`, and putting additional `Device` methods into `Runtime` breaks the single-responsibility principle.

I think we could get rid of `Factory`.  What it's doing is wrapping up the concept of initializing a `Runtime`, so that the associated `Runtime` can use normal constructors and destructors to manage its state -- otherwise, every `Runtime` implementation picks up another bit of state information to indicate whether or not it's been initialized.  To get rid of it, we could instead implement a `Runtime` proxy template that manages the initialization state; I think the functional approach here is easier to maintain in the face of changes to `Runtime`, but on the other hand, some `Runtime` implementations (e.g. LLVM) wouldn't need to be wrapped at all.

I've already talked a bit about `Loader` -- want me to remove it?  I think it's a useful abstraction, but we can ditch it if needed.